File: sdk/java/src/main/java/com/anyvali/schemas/NullSchema.java
package com.anyvali.schemas;

import com.anyvali.IssueCodes;
import com.anyvali.Schema;
import com.anyvali.ValidationContext;

import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Schema that only accepts null.
 */
public class NullSchema extends Schema<Object> {

    public NullSchema() {}

    private NullSchema(NullSchema other) {
        this.coercion = other.coercion;
        this.defaultValue = other.defaultValue;
        this.hasDefault = other.hasDefault;
    }

    @Override
    protected boolean acceptsNull() {
        return true;
    }

    @Override
    protected Object validate(Object input, ValidationContext ctx) {
        if (input != null) {
            String received = anyvaliTypeName(input);
            ctx.addIssue(IssueCodes.INVALID_TYPE,
                    "Expected null, received " + received, "null", received);
            return null;
        }
        return null;
    }

    @Override
    protected Map<String, Object> toNode() {
        return addCommonNodeFields(new LinkedHashMap<>(Map.of("kind", "null")));
    }

    @Override
    protected Schema<Object> copy() {
        return new NullSchema(this);
    }
}

File: sdk/kotlin/src/main/kotlin/com/anyvali/schemas/NullSchema.kt
package com.anyvali.schemas

import com.anyvali.*
import kotlinx.serialization.json.*

class NullSchema : Schema<Nothing?>() {
    override val kind: String = "null"

    fun describe(description: String, opts: DescribeOptions? = null): NullSchema {
        applyDescribe(description, opts)
        return this
    }

    fun metadata(meta: Map<String, Any?>, replace: Boolean = false): NullSchema {
        applyMetadata(meta, replace)
        return this
    }

    override fun validateValue(value: Any?, ctx: ValidationContext): List<ValidationIssue> {
        if (value != null) {
            return listOf(
                ValidationIssue(
                    code = IssueCodes.INVALID_TYPE,
                    path = ctx.path,
                    expected = "null",
                    received = getJsonTypeName(value)
                )
            )
        }
        return emptyList()
    }

    override fun exportNode(): JsonObject = buildJsonObject {
        put("kind", JsonPrimitive("null"))
        addMetadataToNode(this)
    }
}
